### PR TITLE
Create combination_sum_iv.py

### PR DESCRIPTION
**Pull request by kondekarshubham123**
### Describe your change:

Solution added for following question:

You are given an array of distinct integers and you have to tell how many
different ways of selecting the elements from the array are there such that
the sum of chosen elements is equal to the target number tar.

Under hactoberfest event

* [X] Add an algorithm?
* [X] Fix a bug or typo in an existing algorithm?
* [X] Documentation change?

### Checklist:
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [X] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/combination_sum_iv.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+Approach:
+The basic idea is to go over recursively to find the way such that the sum
+of chosen elements is “tar”. For every element, we have two choices
+    1. Include the element in our set of chosen elements.
+    2. Don’t include the element in our set of chosen elements.
+"""
+
+
+from typing import List
```
Comment (CaedenPH) [inline dynamic_programming/combination_sum_iv.py:25]: We target the code in this repo `3.10+` so this isn't necessary. 
You also use `list` in the code anyway
```suggestion
```

File: dynamic_programming/combination_sum_iv.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            return dp_array[target]
+        answer = 0
+        for i in range(len(array)):
+            answer += count_of_possible_combinations_with_dp_array(
+                target - array[i], dp_array
+            )
+        dp_array[target] = answer
+        return answer
+
+    dp_array = [-1 for _ in range(target + 1)]
```
Comment (CaedenPH) [inline dynamic_programming/combination_sum_iv.py:77]:
```suggestion
    dp_array = [-1]  * (target + 1)
```

File: dynamic_programming/combination_sum_iv.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    """
+    Function checks the all possible combinations with using bottom up approach,
+    and returns the count of possible combination in O(N^2) Time Complexity
+    as we are using Dynamic programming array here.
+
+    >>> combination_sum_iv_bottom_up(3, [1,2,5], 5)
+    9
+    """
+
+    dp_array = [0 for _ in range(target + 1)]
```
Comment (CaedenPH) [inline dynamic_programming/combination_sum_iv.py:91]:
```suggestion
    dp_array = [0] * (target + 1)
```

**Review by CaedenPH — COMMENTED** [review]
Looks good!

**Review by CaedenPH — APPROVED** [review]
@cclauss Can we get your review please

File: dynamic_programming/combination_sum_iv.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    def count_of_possible_combinations(target: int) -> int:
+        if target < 0:
+            return 0
+        if target == 0:
+            return 1
+        count = 0
+        for i in range(len(array)):
+            count += count_of_possible_combinations(target - array[i])
+        return count
```
Comment (cclauss) [inline dynamic_programming/combination_sum_iv.py:42]: Please use `sum()` instead.

Comment (kondekarshubham123) [reply dynamic_programming/combination_sum_iv.py:42]: okay, I will do that

Comment (kondekarshubham123) [reply dynamic_programming/combination_sum_iv.py]: Changes made as per requested

File: dynamic_programming/combination_sum_iv.py [hunk 5]
```diff
@@ -36,10 +36,7 @@ def count_of_possible_combinations(target: int) -> int:
             return 0
         if target == 0:
             return 1
-        count = 0
-        for i in range(len(array)):
-            count += count_of_possible_combinations(target - array[i])
-        return count
+        return sum([count_of_possible_combinations(target - each) for each in array])
```
Comment (cclauss) [inline dynamic_programming/combination_sum_iv.py:39]:
```suggestion
        return sum(count_of_possible_combinations(target - item) for item in array)
```

**Review by cclauss — APPROVED** [review]

**Pull request merged**